src/resolution/resolver.ts

```diff
@@ -51,103 +31,86 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
-        } else if (binding.type === BindingTypeEnum.Constructor) {
-            result = binding.implementationType as unknown as T;
-        } else if (binding.type === BindingTypeEnum.DynamicValue && binding.dynamicValue !== null) {
-            result = invokeFactory(
-                "toDynamicValue",
-                binding.serviceIdentifier,
-                () => (binding.dynamicValue as (context: interfaces.Context) => any)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Factory && binding.factory !== null) {
-            result = invokeFactory(
-                "toFactory",
-                binding.serviceIdentifier,
-                () => (binding.factory as interfaces.FactoryCreator<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Provider && binding.provider !== null) {
-            result = invokeFactory(
-                "toProvider",
-                binding.serviceIdentifier,
-                () => (binding.provider as interfaces.Provider<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Instance && binding.implementationType !== null) {
-            result = resolveInstance(
+            break;
+        case "Constructor":
+            result = binding.implementationType as T;
+            break;
+        case "Instance":
+            result = resolveInstance<T>(
                 binding,
-                binding.implementationType,
+                binding.implementationType as interfaces.Newable<T>,
                 childRequests,
-                _resolveRequest(requestScope)
+                _resolveRequest<T>(requestScope)
             );
-        } else {
-            // The user probably created a binding but didn't finish it
-            // e.g. container.bind<T>("Something"); missing BindingToSyntax
-            const serviceIdentifier = getServiceIdentifierAsString(request.serviceIdentifier);
-            throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifier}`);
-        }
+            break;
+        default:
+            result = _resolveFactoryFromBinding(binding,request.parentContext);
+    }
+    return result as T | Promise<T>;
+}
+const _resolveInScope = <T>(
+    requestScope: interfaces.RequestScope,
+    binding:interfaces.Binding<T>,
+    resolveFromBinding: () => T | Promise<T>
+): T | Promise<T> => {
+    let result = _tryGetFromScope(requestScope,binding);
+    if(result !== null){
+        return result;
+    }
+    result = resolveFromBinding();
+    _saveToScope(requestScope,binding, result);
+    return result;
+}
 
+const _resolveBinding = <T>(
+    requestScope: interfaces.RequestScope,
+    request: interfaces.Request,
+    binding:interfaces.Binding<T>,
+): T | Promise<T> => {
+    return _resolveInScope(requestScope,binding, () => {
+        let result = _getResolvedFromBinding<T>(requestScope, request, binding);
```

Comment:
No need to supply the generic
`_getResolvedFromBinding<T>(requestScope, request, binding);`